presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java

```diff
@@ -0,0 +1,163 @@
+        remainingBytesToRevoke -= getMemoryAlreadyBeingRevoked(sqlTasks, memoryPool);
+        requestRevoking(remainingBytesToRevoke, sqlTasks, memoryPool);
+    }
+
+    private long getMemoryAlreadyBeingRevoked(Collection<SqlTask> sqlTasks, MemoryPool memoryPool)
+    {
+        AtomicLong memoryAlreadyBeingRevoked = new AtomicLong();
+        sqlTasks.stream()
+                .filter(task -> task.getTaskInfo().getTaskStatus().getState() == TaskState.RUNNING)
+                .filter(task -> task.getQueryContext().getMemoryPool() == memoryPool)
+                .forEach(task -> task.getQueryContext().accept(new TraversingQueryContextVisitor<Void, Void>()
+                {
+                    @Override
+                    public Void visitOperatorContext(OperatorContext operatorContext, Void context)
+                    {
+                        if (operatorContext.isMemoryRevokingRequested()) {
+                            memoryAlreadyBeingRevoked.addAndGet(operatorContext.getReservedRevocableBytes());
+                        }
+                        return null;
+                    }
+                }, null));
+        return memoryAlreadyBeingRevoked.get();
+    }
+
+    private void requestRevoking(long remainingBytesToRevoke, Collection<SqlTask> sqlTasks, MemoryPool memoryPool)
+    {
+        AtomicLong remainingBytesToRevokeAtomic = new AtomicLong(remainingBytesToRevoke);
+        sqlTasks.stream()
+                .filter(task -> task.getTaskInfo().getTaskStatus().getState() == TaskState.RUNNING)
+                .filter(task -> task.getQueryContext().getMemoryPool() == memoryPool)
+                .sorted(ORDER_BY_CREATE_TIME)
+                .forEach(task -> task.getQueryContext().accept(new TraversingQueryContextVisitor<AtomicLong, Void>()
+                {
+                    @Override
+                    public Void visitQueryContext(QueryContext queryContext, AtomicLong remainingBytesToRevoke)
+                    {
+                        if (remainingBytesToRevoke.get() < 0) {
+                            // exit immediately if no work needs to be done
+                            return null;
+                        }
+                        return super.visitQueryContext(queryContext, remainingBytesToRevoke);
+                    }
+
+                    @Override
+                    public Void visitOperatorContext(OperatorContext operatorContext, AtomicLong remainingBytesToRevoke)
+                    {
+                        if (remainingBytesToRevoke.get() > 0) {
+                            long revokedBytes = operatorContext.requestMemoryRevoking();
+                            if (revokedBytes > 0) {
+                                remainingBytesToRevoke.addAndGet(-revokedBytes);
+                                log.info("(%s)requested revoking %s; remaining %s", memoryPool.getId(), revokedBytes, remainingBytesToRevoke.get());
```

Comment:
@losipiuk will create one :) What kind of stats would you see? How often (count revocation cases), amount?